daemon/api_systems.go

```diff
@@ -272,3 +328,218 @@ func postSystemActionInstall(c *Command, systemLabel string, req *systemActionRe
+	label, errRsp := readFormValue(form, "label")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	testSystem, errRsp := readOptionalFormBoolean(form, "test-system", false)
+	if errRsp != nil {
+		return errRsp
+	}
+
+	markDefault, errRsp := readOptionalFormBoolean(form, "mark-default", false)
+	if errRsp != nil {
+		return errRsp
+	}
+
+	vsetsList, errRsp := readOptionalFormValue(form, "validation-sets", "")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	var splitVSets []string
+	if vsetsList != "" {
+		splitVSets = strings.Split(vsetsList, ",")
+	}
+
+	// TODO: why is this a comma delimited list, rather than multiple values?
+	sequences, err := parseValidationSets(splitVSets)
+	if err != nil {
+		return BadRequest("cannot parse validation sets: %v", err)
+	}
+
+	var snapFiles []*uploadedSnap
+	if len(form.FileRefs["snap"]) > 0 {
+		snaps, errRsp := form.GetSnapFiles()
+		if errRsp != nil {
+			return errRsp
+		}
+
+		snapFiles = snaps
+	}
+
+	batch := asserts.NewBatch(nil)
+	for _, a := range form.Values["assertion"] {
+		if _, err := batch.AddStream(strings.NewReader(a)); err != nil {
+			return BadRequest("cannot decode assertion: %v", err)
+		}
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if err := assertstate.AddBatch(st, batch, &asserts.CommitOptions{Precheck: true}); err != nil {
+		return BadRequest("error committing assertions: %v", err)
+	}
+
+	validationSets, err := assertstate.FetchValidationSets(st, sequences, assertstate.FetchValidationSetsOptions{
+		Offline: true,
+	}, nil)
+	if err != nil {
+		return BadRequest("cannot find validation sets in db: %v", err)
+	}
+
+	slInfo, apiErr := sideloadSnapsInfo(st, snapFiles, sideloadFlags{})
+	if apiErr != nil {
+		return apiErr
+	}
+
+	if len(slInfo.sideInfos) != len(slInfo.tmpPaths) {
+		return InternalError("mismatch between number of snap side infos and temporary paths")
+	}
+
+	localSnaps := make([]devicestate.LocalSnap, 0, len(slInfo.sideInfos))
+	for i := range slInfo.sideInfos {
+		localSnaps = append(localSnaps, devicestate.LocalSnap{
+			SideInfo: slInfo.sideInfos[i],
+			Path:     slInfo.tmpPaths[i],
+		})
+	}
+
+	chg, err := devicestateCreateRecoverySystem(st, label, devicestate.CreateRecoverySystemOptions{
+		ValidationSets: validationSets.Sets(),
+		LocalSnaps:     localSnaps,
+		TestSystem:     testSystem,
+		MarkDefault:    markDefault,
+		// using the form-based API implies that this should be an offline operation
+		Offline: true,
+	})
+	if err != nil {
+		return InternalError("cannot create recovery system %q: %v", label[0], err)
+	}
+
+	ensureStateSoon(st)
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func postSystemActionCreate(c *Command, req *systemActionRequest) Response {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if req.Label == "" {
+		return BadRequest("label must be provided in request body for action %q", req.Action)
+	}
+
+	sequences, err := parseValidationSets(req.ValidationSets)
```

Comment:
It already is a JSON list in the JSON variant of the API. Maybe `parseValidationSets` is a bad name, since it takes a `[]string` and returns a `[]*asserts.AtSequence`.

Reply:
I see, yes the name is slightly confusing